Improve custom tasmod server

PancakeTAS opened the pull request:
- [x] Implement asynchronous server
 - [x] Write packet and authentication system
 - [x] _Start to_ update packets to new server
 - [x] Actually finish updating all packets to new server
 - [x] Write new connection configuration: WONTFIX until game can launch again

PancakeTAS committed:
Add lombok and refactor logger

PancakeTAS committed:
Implement asynchronous client

PancakeTAS committed:
Implement server and update client

PancakeTAS committed:
Implement authentication system

PancakeTAS committed:
Launch server and client on launch

PancakeTAS committed:
Start moving packets to new system

this is painful

PancakeTAS committed:
Continue with packets

PancakeTAS commented:
here is the tasmod server as promised. I've done half of the packets but this process is way to difficult for me cuz I have no idea what I'm doing. I've left you eclipse templates in discord and you can look at the code for examples, it should be self explanatory. You might need to rewrite packets with nbt components to not have nbt components, or wrap a ByteBuf (Unpooled.buffer i believe) around it - at that point just write a util. 

To contribute to this pull request you'll have to pullrequest enhancements/betterserver on my fork.

PancakeTAS commented:
Also FRICK that's a lot of merge conflicts.

PancakeTAS commented:
Quick guide on the packet system:
You can send a message to all clients using `TASmod.server.writeAll`.
You can send a message to the server using `TASmodClient.client.write`.
If you want to send a message to one specific client, you can get the client by it's player uuid using `TASmod.server.getClient` and then use it's `write` method.

For the first two you can use the template "write" and "writeAll" in the eclipse file, it puts a nice error logging block around it for you, and makes it easier to create the packets. It looks like this:
```java
try {
	// packet ${index}: 
	TASmodClient.client.write(ByteBuffer.allocate(${cursor}).putInt(${index}));
} catch (Exception e) {
	TASmod.LOGGER.error("Unable to send packet to server: {}", e);
}
```
This is where you change "index" to a new packet index counting up from the latest one in Client.java.
Then the cursor will jump to allocate(), this is where you calculate how many bytes this packet uses. That would be 4 bytes for an integer (being the id itself) and then add whatever else you need. Pretty simple. Then add a comment to the preplaced comment block describing the packet

To handle the packet go into Client.java and find the serverside/clientside register method and put "handle" in there. It's pretty self explanatory again, it should look like this:
```
// packet ${index}:
this.handlers.put(${index}, buf -> {
	${cursor}
});
```
Put your handling to where the cursor is after entering the index and put your handling stuff in there.

Here you can decide to keep it in the Client class and make it super long... or make a method (NOT A NEW CLASS!) in the class it will act in.

You can see why this is painful lol...

ScribbleTAS commented on src/main/java/com/minecrafttas/server/Client.java:
```diff
@@ -0,0 +1,278 @@
+		);
+		
+		// packet 12: loadstate inputs client
+		this.handlers.put(12, buf -> {
+			try {
+				var nameBytes = new byte[buf.getInt()];
+				buf.get(nameBytes);
+				var name = new String(nameBytes);
+				InputSavestatesHandler.loadstate(name);
+			} catch (Exception e) {
+				TASmod.LOGGER.error("Exception occured during input loadstate: {}", e);
+			}
+		});
+		
+		// packet 13: unload chunks on client
+		this.handlers.put(13, buf ->
+			Minecraft.getMinecraft().addScheduledTask(() -> 
+				SavestatesChunkControl.unloadAllClientChunks()));
+		
+		// packet 14: request client motion
+		this.handlers.put(14, buf -> {
+			var player = Minecraft.getMinecraft().player;
+			if (player != null) {
+				if (!(Minecraft.getMinecraft().currentScreen instanceof GuiSavestateSavingScreen))
+					Minecraft.getMinecraft().displayGuiScreen(new GuiSavestateSavingScreen());
+				
+				try {
+					// packet 15: send client motion to server
+					TASmodClient.client.write(ByteBuffer.allocate(4 + 8+8+8 + 4+4+4 + 1 + 4).putInt(15)
+						.putDouble(player.motionX).putDouble(player.motionY).putDouble(player.motionZ)
+						.putFloat(player.moveForward).putFloat(player.moveVertical).putFloat(player.moveStrafing)
+						.put((byte) (player.isSprinting() ? 1 : 0))
+						.putFloat(player.jumpMovementFactor)
+					);
+				} catch (Exception e) {
+					TASmod.LOGGER.error("Unable to send packet to server: {}", e);
+				}
+			}
+		});
+	}
```
## Declaration
Handlers should be declared in their respective classes via an interface, similar to the current event system. E.g. SavestateNetworkHandler should be registered in SavestateHandler.

If we ever want to reuse the SavestateHandler we can just remove the @ Overrides and have perfectly normal methods that can be hooked up to any other system

Something like `public interface ServerPacketHandler` to declare a handler method. Better than cluttering one file with both.

Something similar to [this](https://github.com/ScribbleTAS/SavestateMod/tree/develop/1.14.4%2B/src/main/java/com/minecrafttas/savestatemod/networking) if you can stand these evil interfaces.

## Identification
Did you know I love advocating for code readability? Like no var and stuff? So having an ID only consist of an int is something I dislike...

We can extend the 2 handler interfaces from a base interface that forces the implementation class to specify an ID... 1.14+ uses these resource locations, why not go basic and use a string...

PancakeTAS replied:
alright so waste two classes which the system was designed to purposefully not have to waste anymore, got it...

yeah sorry I absolutely despise this idea. I designed the system to use anonymous classes/lambdas and you're going to use my really smart implementation but make it super hard to read by adding 2 more classes and defeat it's main purpose too.

I'm generally okay with changing the package "id" to something else but it will still be an integer, period. Feel free to cast an enum around it with a random integer in it's class but DO NOT use strings or anything of that sort.

ScribbleTAS replied:
My goal is to group parts of the code, that do shit for the same thing, **together**. Put everything related to savestates into **one** class, be it events or network handlers. Gets rid of the packet classes, i thought Pancake is happy...

And just so you know it, I am not going for the "least classes possible" way... Just less classes than now... And hey, keep the lambdas if thats possible... Still a good system to register things if its really small, like /folder, just not for the more complicated stuff that i want to keep in the classes where it belongs to. Exactly that mentality of putting different code for different functionality into one class is what makes Lotas 2 so hard to work with... Retracing your steps to get to the code part is just annoying (for the small tacked on features.)... And look Lotas 3 has such a good system. Events and packethandlers are now neatly stored in the mod class. I wanted that system, but using interfaces.

I was also thinking about the enum stuff using it as an id, but didn't write it in the comment

PancakeTAS replied:
then I recommend we do that, but without interfaces (because they simply wouldn't work)
Let's add an annotation purely for readability and use lambdas to call a method in the class directly. That way there doesn't need to be one handler for every single packet

ScribbleTAS replied:
Ok, i can live with that... And write in the javadoc for the annotation where it's registered...

ScribbleTAS commented on src/main/java/com/minecrafttas/tasmod/tickratechanger/TickrateChangerServer.java:
```diff
@@ -165,8 +173,13 @@ public void advanceTick() {
-	}
+    	try {
+	    	// packet 8: advance tick on clients
+	    	TASmod.server.writeAll(ByteBuffer.allocate(4).putInt(8));
```
Write something different than forcing everyone to do maths to allocate the correct amount...

PancakeTAS replied:
yeah I'll admit it, it might not've been the best solution. While it's definitely the fastest it's also supper annoying to use. Here's my solution that is both faster and less memory consuming:
- copy the tas recorders  buffer recycler and fill it with say 1000 buffers of 1kb each.
- then only send the data actually filled (only recently learned about absolute buffer positioning) and recycle the buffer.

As much as I know you LOVE the TAS Recorder, this one class is actually a pretty genius implementation of a thread safe recycler. Basically instead of allocating a new optimal buffer, we reuse existing buffers and note their position and capacity.

on a sidenote my github app crashed 4 times while writing this comment. Why does everyone hate tas recorder :(

ScribbleTAS reviewed (changes requested):
Welp here is my review... I'll have to probably fix this myself, so it's more of a to do list for me... I will definitely rewrite the packets myself so fixing the things I pointed out will probably be enough if anyone feels inclined to fix them...

PancakeTAS commented:
I left my reviews too. Before you go rework all of this I would like to try one more code structuring idea I wanted to for a while now. I'll address both comments in my commits probably tomorrow.

PancakeTAS committed:
Mark fields as final and preinitialize buffers

PancakeTAS committed:
Fix loggers

PancakeTAS commented:
Absolutely no idea if this works, won't be able to test until the game launches again

PancakeTAS committed:
Don't manually calculate sizes

PancakeTAS committed:
Update loom

PancakeTAS committed:
Rewrite to enums

PancakeTAS committed:
Update method calls too

PancakeTAS commented:
your turn

ScribbleTAS committed:
Delombokified code

It's var same thing var Pancake hating ducks. Var makes the code more unreadable
var my opinion. As someone var values readable code var do not approve var this.

ScribbleTAS committed:
Adding packet handler interfaces and helpers for sending packets

This system is a WIP mix between the Pancake method (lambda method) of only lambdas as packet handlers
and my method of adding interfaces to classes for declaring packet handlers (class method).

The class method is good for having one class declare multiple packet handlers,
while the lambda method is better for simple handlers, that do not need a dedicated class.

The lambda method also takes precedence over the class method, meaning you can quickly write packet handlers for debug purposes.

ScribbleTAS commented:
Alright, as it's tradition one person does something, the other rewrites it to their needs.

And since this is mostly my project for once, I will employ some dark arts to make my life easier:

⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⡀⢀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀  
⠀⠀⠀⠀⠀⢀⣠⣴⠾⠟⠛⠛⠙⠛⠻⠷⣦⣄⡀⠀⠀⠀⠀⠀  
⠀⠀⠀⢀⣴⣿⣿⣄⠀⠀⠀⠀⠀⠀⠀⠀⣠⣿⣿⣦⡀⠀⠀⠀  
⠀⠀⣰⡿⠃⠘⣿⡙⠷⣦⣀⠀⠀⢀⣴⠿⢋⣿⠃⠘⢿⣆⠀⠀  
⠀⣰⡿⠁⠀⠀⢹⣇⠀⠈⢙⣷⣾⡛⠁⠀⣼⠏⠀⠀⠈⢿⣇⠀  
⠀⣿⠃⠀⠀⠀⠀⢿⣄⣴⠟⠉⠈⠻⢶⣴⡟⠀⠀⠀⠀⠘⣿⡄  
⢸⡟⠀⠀⠀⠀⣠⡾⣯  Builders ⣿⠿⣦⣀⠀⠀⠀⢿⡇  
⠸⣇⠀⣀⣴⠟⠉⠀⢻⡆⠀⠀⠀⠀⣼⠇⠀⠈⠻⢷⣤⡀⣸⡇  
⠀⣿⡾⠿⠷⠶⠶⠶⠾⣿⠶⠶⠶⢶⡿⠶⠶⠶⠶⠶⠿⣿⣿⠀  
⠀⠘⣿⡄⠀⠀⠀⠀⠀⠸⣇⠀⠀⣾⠃⠀⠀⠀⠀⠀⢠⣿⠃⠀  
⠀⠀⠘⢿⣤⡀⠀⠀⠀⠀⢻⡄⢰⡟⠀⠀⠀⠀⠀⣤⡿⠃⠀⠀  
⠀⠀⠀⠀⠙⢷⣦⣄⠀⠀⠘⣷⣿⠃⠀⠀⣠⣴⡾⠋⠀⠀⠀⠀  
⠀⠀⠀⠀⠀⠀⠈⠙⠻⠷⠶⣿⣿⠶⠿⠟⠋⠁⠀⠀⠀⠀⠀  

Yes you heard that right, I'm using them for constructing Bytbuffers.  
I don't like having to worry about the SecureList and whatnot so I made this *tiny* builder to help with my endeavour.

## Interfaces *and* lamdas for packet handlers?
Thats right, I am trying to make a system that supports both!  
So you have all the advantages in one system and nobody complains...

### Interface method

```java
public class TickSyncClient implements ClientPacketHandler{

	public static final AtomicBoolean shouldTick = new AtomicBoolean(true);
	
	@Override
	public PacketID[] getAcceptedPacketIDs() {
		return new TASmodPackets[] {TASmodPackets.TICKSYNC};
	}

	@Override
	public void onClientPacket(PacketID id, ByteBuffer buf, UUID clientID) {
		TASmodPackets pid = (TASmodPackets) id;

		switch (pid){
			case TICKSYNC:	// Only for demo purposes, you don't need a switch case here, since it's only one packet...
				shouldTick.set(true);
				break;
			default:
				break;
		}
	}
```

Implement ClientPacketHandler and/or ServerPacketHandler, register it in the TASmod init method with `PacketHandlerRegistry.register()` and you are good to go. Supports multiple packethandlers in one method, if you specify more IDs in `getAcceptedPacketIDs()`.

My goal was to group similar packet handlers into one method, instead of spreading all functionality across multiple classes and methods. Hopefully helps with code readability

### Lambda method

```java
public enum TASmodPackets implements PacketID {
	TICKSYNC,
	TICKRATE_SET,
	TICKRATE_ADVANCE,
	SAVESTATE_LOAD,
	SAVESTATE_SAVE,
	/**
	 * <p>CLIENT ONLY.
	 * <p>Opens or closes the savestate screen on the client
	 */
	SAVESTATE_SCREEN(Side.CLIENT, (buf, clientID) -> {
		Minecraft mc = Minecraft.getMinecraft();
		if (!(mc.currentScreen instanceof GuiSavestateSavingScreen))
			mc.displayGuiScreen(new GuiSavestateSavingScreen());
		else
			mc.displayGuiScreen(null);
	});
}
```
Instead of making 2 enums for client and server, I'll only put one enum here and specifying the side.

Also the lambda method takes precedence over the interface method, specifying something in TICKSYNC will overwrite the class I set earlier.

Next up: Fixing all errors

PancakeTAS commented:
> Thats right, I am trying to make a system that supports both!

Selling my implementation as if it was yours huh? noted 😂 

> Yes you heard that right, I'm using them for constructing Bytbuffers.

Oh just make sure to use @Builder from lombok, it does all the work for you ^ﻌ^ฅ♡

ScribbleTAS commented:
> Selling my implementation as if it was yours huh? noted 😂

Right... I've credited you in the commit message and forgot it here... This was more of a comment for you...

ScribbleTAS committed:
WIP Packet rewrite

ScribbleTAS committed:
Added tests for bufferbuilder and server

- Started converting packets

ScribbleTAS committed:
Adding tests for the Client/Server

ScribbleTAS committed:
Continued rewriting packets

ScribbleTAS committed:
Continued with packets

- Switched from ByteBufferBuilder to TASmodBufferbuilder in most cases

ScribbleTAS committed:
Finishing packets, renaming and removing classes

- Added writeNBTTagCompound() to TASmodBufferBuilder

! This commit will start, but crash as soon as you try and join a world !

ScribbleTAS commented:
Alright all errors are fixed and the mod is buildable again...

However now I need to go back and fix all issues with the packets and maybe refactor some networking paths. While I'm at it, I can try to 
do #148
Especially Savestates need a bit of a rewrite

- Ticksync
- Savestates
- Implement Tickratechanger
- Refactor packets so they can also be called in the main menu
- TASmod.getServerInstance() also needs a bit of rethinking.

ScribbleTAS committed:
Increased ttl for ServerTest

ScribbleTAS committed:
Disabled ServerTest for now, doesn't seem to work with github...

ScribbleTAS commented:
Note to self:
Replace `OpenGuiEvents` with a scheduler and have that run when the main menu opens

ScribbleTAS committed:
Added scheduler for opening the main menu

- Reintroduced some classes in SavestateHandlerServer for better visibility

PancakeTAS commented on src/main/java/com/minecrafttas/common/server/ByteBufferBuilder.java:
putting a 163 line builder around a builder to type 3 lines less is something only scribble can do 💀

ScribbleTAS replied:
Only issue is that your 2 lines look like this
```java
var bufIndex = SecureList.POOL.available();
TASmodClient.client.sendToServer(bufIndex, SecureList.POOL.lock(bufIndex).putInt(Client.ServerPackets.TICKRATE_ZERO_TOGGLE.ordinal()).putShort(State.TOGGLE.toShort()));
```
No wonder why you stopped converting the packets...

PancakeTAS replied:
to me this looks perfectly readable

ScribbleTAS replied:
Reading is not the issue, but typing... Writing lots of send requests like that is exhausting imo... Yes eclipse preset bla bla... And having custom types like boolean or NBTTagCompound, that get broken up and added to the buffer in the builder is really handy...

ScribbleTAS replied:
Us mere mortals do not want to wrestle with our brains, when we simply want to send a boolean over the wire... But our godess Pancake has of course no issue typing out 10+ send requests like that...

PancakeTAS replied:
❤️

PancakeTAS commented on src/main/java/com/minecrafttas/common/server/PacketHandlerRegistry.java:
```diff
@@ -0,0 +1,61 @@
+public class PacketHandlerRegistry {
+	private static final List<PacketHandlerBase> REGISTRY = new ArrayList<>();
+
+	public static void register(PacketHandlerBase handler) {
+		if (!REGISTRY.contains(handler)) {
+			REGISTRY.add(handler);
+		} else {
+			System.out.println("Warning...");
+		}
+	}
+
+	public static void unregister(PacketHandlerBase handler) {
+		if (REGISTRY.contains(handler)) {
+			REGISTRY.remove(handler);
+		} else {
+			System.out.println("Warning...");
+		}
+	}
```
best logging ever

ScribbleTAS replied:
Forgot about that... The server was in it's on package and I didn't create a logger for it at the time... Now that I moved it to the common package, I can use the common logger

PancakeTAS commented on src/main/java/com/minecrafttas/common/server/PacketHandlerRegistry.java:
```diff
@@ -0,0 +1,61 @@
+
+		boolean isImplemented = false;
+		for (PacketHandlerBase handler : REGISTRY) {
+			if (ArrayUtils.contains(handler.getAcceptedPacketIDs(), packet)) { 		// TODO Remove the third party library
```
something something Arrays.stream(x).ifPresent() can be used here iirc

ScribbleTAS replied:
You know when you just want it to work and so you just use the library and forget about it...

PancakeTAS replied:
most sane javascript dev 😂

PancakeTAS commented on src/main/java/com/minecrafttas/common/server/exception/InvalidPacketException.java:
```diff
@@ -0,0 +1,17 @@
+
+public class InvalidPacketException extends Exception {
+
+	/**
+	 * 
+	 */
```
wonderful

PancakeTAS commented on src/main/java/com/minecrafttas/common/server/exception/PacketNotImplementedException.java:
```diff
@@ -0,0 +1,22 @@
+public class PacketNotImplementedException extends Exception {
+
+	private static final long serialVersionUID = -8089503724361521594L;
+	
```
you can get rid of the serialversionuids btw, this isn't ever gonna be serialized

PancakeTAS commented on src/main/java/com/minecrafttas/tasmod/networking/TASmodBufferBuilder.java:
```diff
@@ -0,0 +1,212 @@
+			}
+			
+			@Override
+			public void writeBytes(String s) throws IOException {
+				writeString(s);
+			}
+			
+			@Override
+			public void writeByte(int v) throws IOException {
+				buffer.put((byte) v);
+			}
+			
+			@Override
+			public void writeBoolean(boolean v) throws IOException {
+				writeBoolean(v);
+			}
+			
+			@Override
+			public void write(byte[] b, int off, int len) throws IOException {
+				throw new IOException("Not implemented");
+			}
+			
+			@Override
+			public void write(byte[] b) throws IOException {
+				buffer.put(b);
+			}
+			
+			@Override
+			public void write(int b) throws IOException {
+				buffer.put((byte) b);
+			}
+		};
+		
+		try {
+			CompressedStreamTools.write(compound, out);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+		return this;
+	}
```
wtf??
so many better ways to do this for example wrap a Object/DataOutputStream around a ByteArrayOutputStream and write 100 lines less :)

ScribbleTAS replied:
I'm not that versed with streams, so this seemed more logical at the time... I'm more surprised that it even worked like that since I also never worked with anonymous classes...

ScribbleTAS replied:
So uhh how does this work again?
```java
		DataOutputStream stream= new DataOutputStream(new ByteArrayOutputStream());
		
		try {
			CompressedStreamTools.write(compound, stream);
		} catch (IOException e) {
			e.printStackTrace();
		}
		buffer.put(stream);
```
This doesn't work but it was my first thought...

PancakeTAS commented on src/main/java/com/minecrafttas/tasmod/networking/TASmodBufferBuilder.java:
```diff
@@ -0,0 +1,212 @@
+				return TASmodBufferBuilder.readInt(buf);
+			}
+			
+			@Override
+			public void readFully(byte[] b, int off, int len) throws IOException {
+				buf.get(b, off, len);
+			}
+			
+			@Override
+			public void readFully(byte[] b) throws IOException {
+				buf.get(b);
+			}
+			
+			@Override
+			public float readFloat() throws IOException {
+				return TASmodBufferBuilder.readFloat(buf);
+			}
+			
+			@Override
+			public double readDouble() throws IOException {
+				return TASmodBufferBuilder.readDouble(buf);
+			}
+			
+			@Override
+			public char readChar() throws IOException {
+				return buf.getChar();
+			}
+			
+			@Override
+			public byte readByte() throws IOException {
+				return buf.get();
+			}
+			
+			@Override
+			public boolean readBoolean() throws IOException {
+				return TASmodBufferBuilder.readBoolean(buf);
+			}
+		};
+		
+		return CompressedStreamTools.read(input, NBTSizeTracker.INFINITE);
```
same as above except input stream ofc

PancakeTAS commented on src/main/java/com/minecrafttas/tasmod/playback/PlaybackSerialiser.java:
```diff
@@ -98,7 +97,7 @@ public void saveToFileV1(File file, PlaybackController container) throws IOExcep
 	 * @throws IOException When the input container is empty
 	 */
 	public void saveToFileV1Until(File file, PlaybackController container, int index) throws IOException{
-		TASmod.logger.debug(LoggerMarkers.Playback, "Saving playback controller to file {}", file);
+		TASmod.LOGGER.debug(LoggerMarkers.Playback, "Saving playback controller to file {}", file);
 		if (container.size() == 0) {
 			throw new IOException("There are no inputs to save to a file");
 		}
```
just on a sidenote, can we use static imports for loggers? That way we don't need to write TASmod.LOGGER every time.

`import static com.minecrafttas.tasmod.TASmod.LOGGER;`

ScribbleTAS replied:
Sure

PancakeTAS commented on src/main/java/com/minecrafttas/tasmod/util/Scheduler.java:
```diff
@@ -4,28 +4,28 @@
- * Schedules a lambda to be run in the next tick
+ * A simple scheduling interface
  * 
  * @author Scribble
  *
  */
-public class TickScheduler {
+public class Scheduler {
 	
-	Queue<TickTask> queue = new ConcurrentLinkedQueue<>();
+	Queue<Task> queue = new ConcurrentLinkedQueue<>();
 	
 	public void runAllTasks() {
-		TickTask task;
+		Task task;
 		while((task = queue.poll()) != null) {
 			task.runTask();
 		}
 	}
 	
-	public void add(TickTask task) {
+	public void add(Task task) {
 		queue.add(task);
 	}
 	
 	@FunctionalInterface
-	public interface TickTask {
+	public interface Task {
 		public void runTask();
 	}
 }
```
unrelated but check out Executors.newXXExecutor() in the future, could be nice for easy multithreading

PancakeTAS commented on src/test/java/common/server/ByteBufferBuilderTest.java:
tests? documentation? we don't document tests? The person fixing it has absolutely no idea why X behavior should happen ._.

ScribbleTAS replied:
Doing that at the end... Famous last words...

PancakeTAS reviewed (commented):
please work on your documentation 🙏

ScribbleTAS committed:
Added logger as static import

ScribbleTAS committed:
Changed default port, added static import for logger

- Added *proper* logging in PacketHandlerRegistry
- Removed serial verion id

ScribbleTAS commented:
*Note to self:*  
Implement tickratechanger packets

ScribbleTAS committed:
Implementing tickrate packets, registering packet handlers

ScribbleTAS commented:
Welcome to another episode of

# Scribbles cursed code snippets

In todays episode we have:

## Switch case statements *inside* switch case statements

```java
	public void onClientPacket(PacketID id, ByteBuffer buf, UUID clientID) {
		TASmodPackets packet = (TASmodPackets) id;
		
		switch (packet) {
		case TICKRATE_CHANGE:
			float tickrate = TASmodBufferBuilder.readFloat(buf);
			changeClientTickrate(tickrate);
			break;
		case TICKRATE_ADVANCE:
			advanceClientTick();
			break;
		case TICKRATE_ZERO:
			TickratePauseState state = TASmodBufferBuilder.readTickratePauseState(buf);

			switch (state) {
			case PAUSE:
				pauseClientGame(true);
				break;
			case UNPAUSE:
				pauseClientGame(false);
				break;
			case TOGGLE:
				togglePauseClient();
			default:
				break;
			}
			break;

		default:
			throw new PacketNotImplementedException(packet, this.getClass());
		}
	}
```

PancakeTAS commented:
please indent the case and default one further from the switch itself 🙏 
Also if I see one undocumented switch case I will fire you

ScribbleTAS commented:
> please indent the case and default one further from the switch itself 🙏 Also if I see one undocumented switch case I will fire you

But the Enums are documented ._. And funny enough, this is the default eclipse indentation... I just did Ctrl+F on it and eclipse doesn't indent it...

PancakeTAS commented:
https://github.com/PancakeTAS/BackgroundLed/blob/develop/arduino.ino

i want this kind of documentation /j

ScribbleTAS committed:
Documentation and small code improvements

PancakeTAS reviewed (commented):
no custom exceptions.

ScribbleTAS commented:
Wut?
Does that mean remove all custom exceptions?
Should I add more?
Earlier you were completely fine with them, pointing out the serial version uuid...
And I want your reason as to why... Too many classes again?
They have helped me already, having a contructor for ypur messages is really nice...

The custom exceptions will stay... I also have some for savestatemod and others

PancakeTAS commented:
> Should I add more?

*wtf lol*

Nah I recently changed my mind *again*. I don't like custom exceptions because they have no use other than to save a single sentence comment. If they would provide different data in the stacktrace sure, but these are just... eh

ScribbleTAS commented:
> Nah I recently changed my mind _again_. I don't like custom exceptions because they have no use other than to save a single sentence comment. If they would provide different data in the stacktrace sure, but these are just... eh

Now I'm intrigued of how to provide different data in the stacktrace, because for networking this does sound useful...

ScribbleTAS commented:
> wtf??
> so many better ways to do this for example wrap a Object/DataOutputStream around a ByteArrayOutputStream and write 100  
> lines less :)

@PancakeTAS Alright what do you mean with that... Before I do that wrong again, I'll just ask you. Because I admit, I don't like my solution either, but so far nothing else worked

PancakeTAS commented:
i will take a look today... if I don't forget

ScribbleTAS committed:
Fixed requesting motion, simplified NBTTagCompound sending

ScribbleTAS committed:
Fixing and discovering savestating issues

ScribbleTAS commented:
Right, now the issue is that the uuid we have in the main menu is not necessarily the uuid that we have on the server... So we may connect to the custom server with a different uuid than to the vanilla server, which throws things off...

PancakeTAS commented:
Which UUID?

ScribbleTAS commented:
So, the custom server starts in the main menu and "authenticates" with the session uuid... for indev it uses a random uuid. Once you join a minecraft server, the entityplayermp gets a random uuid as well... So when i try to send something to a client for savestates, i can't because the player uuid is not the same as the client uuid

ScribbleTAS committed:
Added defaults to infohud, added reset button, switched middleclick and rightclick

ScribbleTAS committed:
Changing to usernames instead off uuid, fixing BufferUnderflow crashes

ScribbleTAS commented:
*Note to self:*  
- [x] Move TASStateServer to PlaybackControllerServer
- [x] Reorder EventHandler and PacketHandler registering in TASmod.class and TASmodClient.class to be split

ScribbleTAS committed:
Rearranging registering in main methods and documentation 🚂

ScribbleTAS commented:
> Write new connection configuration: WONTFIX until game can launch again

@PancakeTAS What do you mean by that?

PancakeTAS commented:
well what we discussed - connecting to the server at game launch using a configuration file

ScribbleTAS commented:
Ah so it was that

ScribbleTAS committed:
Fixed labels going off screen in InfoHud

ScribbleTAS committed:
Networking for PlaybackControllerClient and TASstateServer to PlaybackControllerServer

ScribbleTAS commented:
# Testing
Here is my todo list for testing everything. Also add trace logging

## Server
- [x] TickSyncServer
- PlaybackControllerServer
	- [x] Statesync
	- [x] StatesyncInitial (Needs multiplayer server)
- SavestateHandlerServer
	- [x] Savestate
	- [x] Loadstate

- TickrateChangerServer
	- [x] Change
	- [x] Unpause
	- [x] Toggle
- KillTheRNGHandler
	- [x] KilltheRNG
## Client
- PlaybackControllerClient
	- [x] Save
	- [x] Load
	- [x] Play
	- [x] Record
	- [x] Fullplay
	- [x] Fullrecord
	- [x] Restartandplay
	- [x] Playuntil
	- [x] Clearinputs
	- [x] Teleport

- SavestateHandlerClient
	- [x] PlayerMotion
	- [x] Screen
	- [x] Unloadchunks

ScribbleTAS committed:
More documentation for packets

- Fixed crash when client motion tag does not exist

ScribbleTAS committed:
Updating loader version to 0.14.23

ScribbleTAS committed:
Adding log4j config to server console

ScribbleTAS committed:
Testing multiplayer connections

ScribbleTAS commented:
Todo Prevent sending packets, when no connection is open

ScribbleTAS committed:
Rewriting connection things

ScribbleTAS committed:
Fixing server closing

- Fixed local server closing when leaving the singeplayer world
- Fixed client being unresponsive in main menu when not connected to a custom server
- Various logging changes
- Add testing keybinds for connecting and disconnecting from the custom server

ScribbleTAS committed:
Fixed connecting to the wrong server

Crash fix

ScribbleTAS committed:
Fixed tickadvance not working as intended

PancakeTAS commented on src/main/java/com/minecrafttas/common/server/Client.java:
```diff
@@ -78,6 +81,12 @@ public Client(String host, int port, PacketID[] packetIDs, String name) throws E
 	 */
 	public Client(AsynchronousSocketChannel socket, PacketID[] packetIDs) {
 		this.socket = socket;
+		try {
+			this.socket.setOption(StandardSocketOptions.SO_KEEPALIVE, true);
+			this.socket.setOption(StandardSocketOptions.TCP_NODELAY, true);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
```
just add throws and do proper error handling?

ScribbleTAS replied:
Sure

PancakeTAS commented on src/main/java/com/minecrafttas/common/server/Server.java:
```diff
@@ -45,19 +45,10 @@ public Server(int port, PacketID[] packetIDs) throws Exception {
-					}
-				}
-			}
-
```
for the future,
`clients.removeIf(Socket::isClosed);`
is a pretty neat way of doing this

PancakeTAS reviewed (commented):
i just realized I clicked the wrong button, anyways looks good

ScribbleTAS commented:
I should invest some time in a proper disconnect system that checks when a socket is closed... Also a disconnect packet would be quite neat I think.

ScribbleTAS committed:
Added disconnect packet and check for timeout

- Fixed EventOpenGuiScreen not firing

ScribbleTAS committed:
Fixed Redirect throwing errors

ScribbleTAS commented:
@PancakeTAS I added a timeout system to this mess that currently checks every 10 seconds, if a client is still alive... Maybe I'll make that tick based, so that tickrate 0 and low tickrates will not cause timeouts...

I made an onClose callback that removes clients from the list of clients in Server.java... It's also for Events and whatnot...

And disconnect() actually sends a packet to the server now that tells the server to close...
Tested it and disconnecting and connecting seems to be working!

Maybe you can take a look at this rq as the connectivity expert

PancakeTAS reviewed (commented):
so instead of using TCPs builtin timeout system, you wrote your own... well as long as it works I don't mind

ScribbleTAS commented:
From what I have gathered, TCP has no timeout system... Else I wouldn't have so many problems...

ScribbleTAS committed:
Fixed client and server timing out in tickrate 0

ScribbleTAS committed:
Fixed gui controls crashing in tr 0

ScribbleTAS committed:
Updated junit

ScribbleTAS commented:
Note to self:
Remove ticksync connection when in main menu and connected to singleplayer

Add keepalive packet...

ScribbleTAS committed:
Add property for "integrated" server

ScribbleTAS committed:
Removed timeout system

ScribbleTAS commented:
Note to self:
Check out Ticksync and when it's enabled again to see if it's correct. Don't forget that!

Afterwards implement SyncStatePacket... When sending that to the server I got an error idk why...

ScribbleTAS commented:
Welcome to Scribble losing his mind again doing very dumb mistakes.  
I'm getting this error:
```
[14:34:11][Thread-19/ERROR] (Common/THROWING[ EXCEPTION ]) Throwing
java.nio.BufferUnderflowException: null
	at java.nio.Buffer.nextGetIndex(Buffer.java:510) ~[?:1.8.0_372]
	at java.nio.HeapByteBuffer.getShort(HeapByteBuffer.java:323) ~[?:1.8.0_372]
	at com.minecrafttas.tasmod.networking.TASmodBufferBuilder.readTASState(TASmodBufferBuilder.java:81) ~[main/:?]
	at com.minecrafttas.tasmod.playback.PlaybackControllerServer.onServerPacket(PlaybackControllerServer.java:73) ~[main/:?]
```
Great. This tells me that the packet I am sending inbetween the server and client has no data and it fails to read that data.

So you start checking every instance where you send that particular packet. Nope, everything is correct here.  
Then you notice that when you join singleplayer it works. But if you quit to menu and go to singleplayer again, then it fails.

Hmmm... Then you find out that the packet is sent twice? Then you check the connection. Nope only one connection.  
Then you finally find the issue:

```java
public void onServerInit(MinecraftServer server) {
	playbackControllerServer=new PlaybackControllerServer();
	PacketHandlerRegistry.register(playbackControllerServer);
}
```
So everytime I join singleplayer, I register a new packethandler.  
I am checking if that object is already registered in the registry
```java
if (!REGISTRY.contains(handler)) {
	REGISTRY.add(handler);
} else {
	Common.LOGGER.warn("Trying to register packet handler {}, but it is already registered!", handler.getClass().getName());
}
```
but due to the `new` keyword, this is a new object so it happily registers 2 instances of the same class

PancakeTAS commented:
buffer underflow exception doesn't necessarily mean there's no data at all, it just means that in the buffer, the `position` has exceeded the `limit` of the buffer. In other words, you read everything from the buffer and now it is empty. That's why you see me call flip() (or something like that) when reading or sending a packet, it resets the position and puts the limit to the current position. Naturally if you try to read the same buffer twice without calling flip, you'll underflow the buffer (though idk why it's called underflow).

Anyways I'll use your small little hiccup as an argument for why registries are a terrible idea :3

By the way, I'm pretty sure the contains() method called equals() on every object, to see if they are equal. If you want to prevent bugs like this in the future, you can override that method in the class and implement your own is equal check.

ScribbleTAS commented:
> I'm pretty sure the contains() method called equals() on every object

Nice idea, however, the object that I'm registering is an interface, meaning I'd have to override all implementations or do some abstract class bs to get this done the same way.

Instead I opted for this
```java
private static boolean containsClass(PacketHandlerBase handler) {
	for(PacketHandlerBase packethandler : REGISTRY) {
		if(packethandler.getClass().equals(handler.getClass())) {
			return true;
		}
	}
	return false;
}
```
And low and behold, look I've done the same thing twice
![image](https://github.com/MinecraftTAS/TASmod/assets/28013575/f64af715-0f68-4eb8-af52-abe50ad83de2)

ScribbleTAS committed:
Added a trace statement for network packets and it's contents

- Added an option to packets (In this case TASmodPackets) to be turned off for this trace view

ScribbleTAS committed:
Networking fixes

- Fixed the ability to register 2 instances of packet handlers with the same class
- Removed STATESYNC_INITIAL
- [Savestates] Fixed unexpected behaviour
- [PlaybackController] Fixed errors in console

ScribbleTAS committed:
[Savestates] Fixed a rare crash with gui screens

ScribbleTAS committed:
Trying to fix savestates

ScribbleTAS commented:
Note to self:
- Make firing events better.
- Savestates take 3 ticks to load? -> Let the world tick?
- You can't unpause from tickrate 0 in gui screens?

PancakeTAS reviewed (commented):
idk what I just reviewed, there were only deletions hehe

ScribbleTAS committed:
[Savestates] Remove WASLOADING state

ScribbleTAS committed:
[Event] Drafting new event firing

ScribbleTAS commented on src/main/java/com/minecrafttas/common/events/EventListenerRegistry.java:
```diff
@@ -0,0 +1,83 @@
+			Method methodToFire = null;
+			
+			for(Method method : eventClass.getMethods()) {
+				if(Modifier.isAbstract(method.getModifiers())) {
+					methodToFire = method;
+					break;
+				}
+			}
+			
+			for(EventListenerRegistry.EventBase eventListener : EventListenerRegistry.EVENTLISTENER_REGISTRY) {
+				Class<?>[] interfaces = eventListener.getClass().getInterfaces();
+				for(Class<?> interfaze : interfaces) {
+					if(interfaze.equals(eventClass)) {
+						Method methodInClass = null;
+							Method[] methods = eventListener.getClass().getMethods();
+							for(Method method : methods) {
+								if(method.getName().equals(methodToFire.getName())) {
+									methodInClass = method;
+									break;
+								}
+							}
+						
+						try {
+							return methodInClass.invoke(eventListener, eventParams);
+						} catch (IllegalAccessException e) {
+							e.printStackTrace();
+							return null;
+						} catch (IllegalArgumentException e) {
+							e.printStackTrace();
+							return null;
+						} catch (InvocationTargetException e) {
+							e.printStackTrace();
+							return null;
+						}
+					}
+				}
+			}
+			return null;
+		}
+	}
```
@PancakeTAS Here you go, more cursed code...

PancakeTAS replied:
my only request is that you merge the catches into one Exception

ScribbleTAS committed:
Revert "[Event] Drafting new event firing"

This reverts commit 8614dbb4ed0b467586e7b7d572d20edfdb4c4111.

ScribbleTAS committed:
[Networking] Added config option for automatically connecting to a custom server on startup

ScribbleTAS committed:
[Networking] Documentation fixes 🚂

ScribbleTAS commented:
@PancakeTAS How many months did this take? idk, but pls review

ScribbleTAS committed:
[Networking] Finishing documentation 🚂 for packet ids
- Fixed clearinputs packet being sent twice
- Renamed SYNCSTATE packet to PLAYBACK_STATE
- Renamed CLEAR_INPUTS to PLAYBACK_CLEAR_INPUTS

PancakeTAS reviewed (commented):
approved
